tests/si/rendering.cc

```diff
@@ -49,30 +121,324 @@ APP("ui rendering")
     auto ps_clear = ctx.make_shader(shader_code_clear, "main_ps", pr::shader::pixel);
+    auto vs_ui = ctx.make_shader(shader_code_ui, "main_vs", pr::shader::vertex);
+    auto ps_ui = ctx.make_shader(shader_code_ui, "main_ps", pr::shader::pixel);
+    auto vs_ui_wireframe = ctx.make_shader(shader_code_ui_wireframe, "main_vs", pr::shader::vertex);
+    auto ps_ui_wireframe = ctx.make_shader(shader_code_ui_wireframe, "main_ps", pr::shader::pixel);
+
+    using vertex_t = si::Default2DMerger::vertex;
+    phi::vertex_attribute_info vert_attrs[] = {phi::vertex_attribute_info{"POSITION", unsigned(offsetof(vertex_t, pos)), phi::format::rg32f},
+                                               phi::vertex_attribute_info{"TEXCOORD", unsigned(offsetof(vertex_t, uv)), phi::format::rg32f},
+                                               phi::vertex_attribute_info{"COLOR", unsigned(offsetof(vertex_t, color)), phi::format::rgba8un}};
+
+    auto fb_info = pr::framebuffer_info().target(ctx.get_backbuffer_format(swapchain), pr::blend_state::alpha_blending());
+    auto gp_info = pr::graphics_pass(vs_ui, ps_ui).arg(1, 0, 1, true).vertex(sizeof(vertex_t), vert_attrs);
+    auto pso_ui = ctx.make_pipeline_state(gp_info, fb_info);
+
+    auto fb_info_wf = pr::framebuffer_info().target(ctx.get_backbuffer_format(swapchain));
+    auto gp_info_wf = pr::graphics_pass(vs_ui_wireframe, ps_ui_wireframe).arg(0, 0, 0, true).vertex(sizeof(vertex_t), vert_attrs).wireframe();
+    auto pso_ui_wf = ctx.make_pipeline_state(gp_info_wf, fb_info_wf);
 
     si::gui ui;
+    si::Default2DMerger ui_merger;
+
+    // init input
+    inc::da::input_manager input;
+    input.initialize(100);
+    auto is_in_text_edit = false;
+
+    enum e_input : uint64_t
+    {
+        action_mouse_x = 1000,
+        action_mouse_y,
+        action_mouse_left,
+    };
+
+    input.bindMouseAxis(action_mouse_x, 0);
+    input.bindMouseAxis(action_mouse_y, 1);
+    input.bindMouseButton(action_mouse_left, inc::da::mouse_button::mb_left);
+
+    // upload font data
+    // TODO: mipmaps
+    auto const& font = ui_merger.get_font_atlas();
+    auto font_tex = ctx.make_texture({font.width, font.height}, pr::format::r8un, 1);
+    auto font_arg = pr::argument();
+    font_arg.add(font_tex);
+    font_arg.add_sampler(pr::sampler_filter::min_mag_mip_linear, 0, pr::sampler_address_mode::clamp);
+    {
+        auto fb = ctx.make_upload_buffer(font.data.size());
+        CC_ASSERT(int(font.data.size()) == font.width * font.height);
+        ctx.write_to_buffer_raw(fb, font.data);
+
+        auto frame = ctx.make_frame();
+        // needed?
+        // frame.transition(fb, pr::state::copy_src);
+        // frame.transition(font_tex, pr::state::copy_dest);
+        frame.copy(fb, font_tex);
+        frame.transition(font_tex, pr::state::shader_resource, pr::shader::pixel);
+        ctx.submit(cc::move(frame));
+        ctx.flush(); // needed?
```

Comment:
this is necessary here because `fb` from line 169 will run out of scope and get destroyed despite being in flight. 
however, it's an unnecessary stall, this can now be solved more elegantly by calling `frame.free_deferred_after_submit` (on `fb` after unlocking it)

`ctx.free_deferred`: frees the resource as soon as the GPU has reached current CPU epoch
`frame.free_deferred_after_submit`: calls the function above once the frame is submitted, meaning the resource can still be used by the frame itself.

Reply:
would it make sense to add an `auto_buffer&&` overload to free_deferred_after_submit?
imho it feels/reads a bit more natural to do:
```
frame.free_deferred_after_submit(cc::move(fb));
```
instead of
```
frame.free_deferred_after_submit(fb.unlock());
```